mongo/with_transactions_test.go

```diff
@@ -259,6 +260,41 @@ func TestConvenientTransactions(t *testing.T) {
 		assert.True(t, ok, "expected context for abortTransaction to contain ctxKey")
 		assert.Equal(t, "foobar", ctxValue, "expected value for ctxKey to be 'world', got %s", ctxValue)
 	})
+	t.Run("commitTransaction timeout allows abortTransaction", func(t *testing.T) {
+		coll := db.Collection("test")
+
+		session, err := client.StartSession()
+		assert.Nil(t, err, "StartSession error: %v", err)
+		defer session.EndSession(context.Background())
+
+		err = WithSession(context.Background(), session, func(sessionContext SessionContext) error {
+			// Start transaction
+			err = session.StartTransaction()
+			assert.Nil(t, err, "StartTransaction error: %v", err)
+
+			// Insert a document
+			_, err := coll.InsertOne(sessionContext, bson.D{
+				{"ID", primitive.NewObjectID()},
+				{"val", 17},
+			})
+			assert.Nil(t, err, "InsertOne error: %v", err)
+
+			// Set a timeout of 0 for commitTransaction
+			commitTimeoutCtx, commitCancel := context.WithTimeout(sessionContext, 0)
+			defer commitCancel()
+
+			// commitTransaction results in context.DeadlineExceeded
+			commitErr := session.CommitTransaction(commitTimeoutCtx)
+			assert.True(t, errors.Is(commitErr, context.DeadlineExceeded),
+				"expected context.DeadlineExceeded error; got %v", commitErr)
+
+			// abortTransaction without error
+			abortErr := session.AbortTransaction(context.Background())
+			assert.Nil(t, abortErr, "AbortTransaction error: %v", abortErr)
+
```

Comment:
Ideally, I would also check `!session.clientSession.TransactionCommitted()`, but that's not a public API; I could use the unstable `XSession` interface if we really want to. 

I could also make sure that the document is not actually inserted to the database.

Reply:
Notably, this was necessary for the unified test runner to check transaction state:
https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/unified/testrunner_operation_test.go#L80:L80

In general, I prefer not relying on private API for tests unless it is necessary. Then test code becomes a consumer, and it makes it more difficult to change the private API. But, there is precedent. If you feel it adds to the coverage of the test it seem ok in this case to add a similar assertion.

Reply:
Since there's precedent, I've added in a check for `!TransactionCommitted()` that uses private API. The main issue we're trying to solve is that session state is moved to "Committed" when it shouldn't be, so the check seems integral.

Reply:
Sounds good to me.